File: Data Structures & Algorithms/count-number-of-islands/submission-0.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return -1

        dirs = [[0,1], [0,-1], [1,0], [-1,0]]
        rows, cols = len(grid), len(grid[0])

        def bfs(row, col):
            queue = deque([(row, col)])
            grid[row][col] = '0'

            while queue:
                curr_row, curr_col = queue.popleft()
                for row_add, col_add in dirs:
                    nr, nc = curr_row + row_add, curr_col + col_add
                    if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == '1':
                        queue.append((nr, nc))
                        grid[nr][nc] = '0'

        count = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1':
                    count += 1
                    bfs(r, c)
        return count

File: Data Structures & Algorithms/islands-and-treasure/submission-0.java
class Solution {
    public void islandsAndTreasure(int[][] grid) {
        // edge case
        if (grid == null) {
            return;
        }

        // get the rows and the cols
        int rows = grid.length;
        int cols = grid[0].length;

        // create the queue
        Queue<int[]> queue = new LinkedList<>();

        // iterate over the grid and collect all of the starting points (treasure chest coordinates)
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == 0) {
                    queue.add(new int[] {r, c});
                }
            }
        }

        // check if the queue is empty (no treasure chests)
        if (queue.isEmpty()) {
            return;
        } 

        // directions
        int[][] dirs = {{0,1}, {0, -1}, {1,0}, {-1,0}};

        // iterate over the queue
        while (!queue.isEmpty()) {
            // get the current element
            int[] curr = queue.poll();
            int row = curr[0];
            int col = curr[1];

            // iterate over the directions
            for (int[] dir: dirs) {
                int newRow = row + dir[0];
                int newCol = col + dir[1];

                if (newRow < 0 || newRow == rows || newCol < 0 || newCol == cols || grid[newRow][newCol] != Integer.MAX_VALUE) {
                    continue;
                }

                queue.add(new int[] {newRow, newCol});                
                grid[newRow][newCol] = grid[row][col] + 1;      // this will ensure that we are not changing the value of the treasure chest cell
            }

        }


    }
}
